docs/data_model.md

```diff
@@ -0,0 +1,591 @@
+
+```mermaid
+erDiagram
+    Contract }|--|| Statement : ""
+    Contract }|--|| ContractTemplate : ""
+
+    Statement {
+        uuid id
+        enum month
+        integer year
+        date deadline_date
+        uuid cohort_id
+        uuid lead_provider_id
+        datetime marked_as_paid_at
+        decimal reconcile_amount
+        string state
+    }
+
+    Contract {
+        uuid statement_id
+        uuid course_id
+        uuid contract_template_id
+    }
+
+    ContractTemplate {
+        boolean special_course
+        decimal recruitment_target
+        decimal per_participant
+        decimal output_payment_percentage
+        decimal number_of_payment_periods
+        decimal service_fee_percentage
+        decimal service_fee_installments
+    }
+```
+
+### 4.5. **User and Provider Management**
+
+   - **User ID Management**: `ParticipantIdChange` enables the tracking of changes in user (`User`) identifiers, essential for maintaining a clear user history.
+   - **API Access**: `ApiToken` allows `LeadProvider`s secure access to the system via hashed tokens, facilitating data access for providers while ensuring security.
+
+   These auxiliary processes help maintain user data integrity and allow providers secure access to manage applications and declarations.
+
+```mermaid
+erDiagram
+    ParticipantIdChange }|--|| User : ""
+    LeadProvider }|--|| ApiToken : ""
+
+    User {
+        uuid id
+        string email
+        string full_name
+        string teacher_reference_number
+        datetime updated_at
+    }
+
+    LeadProvider {
+        uuid id
+        string name
+    }
+
+    ParticipantIdChange {
+        uuid id
+        uuid user_id
+        uuid from_participant_id
+        uuid to_participant_id
+    }
+
+    ApiToken {
+        uuid id
+        uuid lead_provider_id
+        string hashed_token
+        datetime last_used_at
+    }
+```
+
+#### Key Data Flow Summary:
+
+1. **Application Creation** ➔ Triggers cohort and schedule associations for user applications.
+2. **Declaration Submission** ➔ Captures user progress and outcomes, feeding into statements for financial tracking.
+3. **Statement and Contract Processing** ➔ Ensures compliance with financial terms for providers, courses, and user declarations.
+4. **Provider and User Management** ➔ Supports secure provider access and accurate user tracking across application processes.
+
+### Overall Workflow
+
+1. **User applies** for a course through an `Application`, selecting relevant schedules and providers.
+2. **Declarations** are made based on user progress, feeding into statements and tracking outcomes.
+3. **Statements** and **Contracts** manage finances, ensuring terms are met.
+4. **Providers** access data through `ApiToken`, while user identifiers are tracked to ensure data accuracy.
+
+## 5. Data Integrity and Constraints
```

Comment:
can we remove this section doesn't feel relevant also our primary keys are integers right now